consensus/src/firststep/handler.rs

```diff
@@ -106,65 +106,95 @@ impl<D: Database> MsgHandler<Message> for Reduction<D> {
-                    SvType::FirstReduction,
-                ) {
-                    return Ok(HandleMsgOutput::FinalResult(m));
+            // Record result in global round registry
+            _ = self.sv_registry.lock().await.add_step_votes(
+                step,
+                hash,
+                sv,
+                SvType::FirstReduction,
+                quorum_reached,
+            );
+
+            if quorum_reached {
+                // if the votes converged for an empty hash we invoke halt
+                if hash == [0u8; 32] {
+                    tracing::warn!("votes converged for an empty hash");
+
+                    return Ok(final_result_with_timeout(
+                        StepVotes::default(),
+                        ledger::Block::default(),
+                    ));
                 }
 
-                // If step is different from the current one, this means
-                // * collect * func is being called from different iteration
-                if step != self.curr_step {
-                    return Ok(HandleMsgOutput::Result(msg));
+                if hash != self.candidate.header().hash {
+                    // If the block generator is behind this node, we'll miss
+                    // the candidate block.
+                    if let Ok(block) = self
+                        .db
+                        .lock()
+                        .await
+                        .get_candidate_block_by_hash(&hash)
+                        .await
+                    {
+                        return Ok(final_result(sv, block));
+                    }
+
+                    tracing::error!("Failed to retrieve candidate block.");
+                    return Ok(empty_result!());
                 }
-            }
 
-            // if the votes converged for an empty hash we invoke halt
-            if hash == [0u8; 32] {
-                tracing::warn!("votes converged for an empty hash");
-
-                return Ok(final_result_with_timeout(
-                    StepVotes::default(),
-                    ledger::Block::default(),
-                ));
+                return Ok(final_result(sv, self.candidate.clone()));
             }
+        }
 
-            if hash != self.candidate.header().hash {
-                // If the block generator is behind this node, we'll miss the
-                // candidate block.
-                if let Ok(block) = self
-                    .db
-                    .lock()
-                    .await
-                    .get_candidate_block_by_hash(&hash)
-                    .await
-                {
-                    return Ok(final_result(sv, block));
-                }
+        Ok(HandleMsgOutput::Pending(msg))
+    }
 
-                tracing::error!("Failed to retrieve candidate block.");
-                return Ok(empty_result!());
-            }
+    /// Collects the reduction message from former iteration.
+    async fn collect_from_past(
```

Comment:
I'm a bit confused by these `collect_from_past` functions. 
Shouldn't this be part of #1098 ? What's being changed here? 
Are we collecting more Reduction votes than before?

Can you clarify it?

Reply:
Before, there was a single function `collect` responsible for handling reduction msg from both the current step and a former step. This design led to confusion, respectively bugs. To simplify this,  the `collect` function now exclusively processes only messages from the current step while a new functoin `collect_from_past` is dedicated for handling prev-steps msgs. 
In short, this is considered a refactoring change.